packages/opentelemetry-exporter-datadog/src/datadogSpanProcessor.ts

```diff
@@ -0,0 +1,203 @@
+ *
+ * Only spans that are sampled are converted.
+ */
+export class DatadogSpanProcessor implements SpanProcessor {
+  private readonly _bufferTimeout: number;
+  private readonly _maxQueueSize: number;
+  private readonly _maxTraceSize: number;
+  private _isShutdown = false;
+  // private _exporter: SpanExporter;
+  private _timer: NodeJS.Timeout | undefined;
+  private _traces = new Map();
+  private _traces_spans_started = new Map();
+  private _traces_spans_finished = new Map();
+  private _check_traces_queue: Set<string> = new Set();
+  public readonly logger: Logger;
+
+  constructor(
+    private readonly _exporter: SpanExporter,
+    config?: DatadogBufferConfig
+  ) {
+    this.logger = (config && config.logger) || new NoopLogger();
+    this._bufferTimeout =
+      config && typeof config.bufferTimeout === 'number'
+        ? config.bufferTimeout
+        : DEFAULT_BUFFER_TIMEOUT_MS;
+
+    this._maxQueueSize =
+      config && typeof config.maxQueueSize === 'number'
+        ? config.maxQueueSize
+        : DEFAULT_MAX_QUEUE_SIZE;
+    this._maxTraceSize =
+      config && typeof config.maxTraceSize === 'number'
+        ? config.maxTraceSize
+        : DEFAULT_MAX_TRACE_SIZE;
+  }
+
+  forceFlush(): void {
+    if (this._isShutdown) {
+      return;
+    }
+    this._flush();
+  }
+
+  shutdown(): void {
+    if (this._isShutdown) {
+      return;
+    }
+    this.forceFlush();
+    this._isShutdown = true;
+    this._exporter.shutdown();
+  }
+
+  // does nothing.
+  onStart(span: ReadableSpan): void {
+    if (this._isShutdown) {
+      return;
+    }
+
+    if (this._allSpansCount(this._traces_spans_started) >= this._maxQueueSize) {
+      this.logger.debug('Max Queue Size reached, dropping span');
+      return;
+    }
+
+    const traceId = span.spanContext['traceId'];
+
+    if (!this._traces.has(traceId)) {
+      this._traces.set(traceId, []);
+      // this._traces.get(traceId).set(SPANS_KEY, [])
+      this._traces_spans_started.set(traceId, 0);
+      this._traces_spans_finished.set(traceId, 0);
+    }
+
+    const trace = this._traces.get(traceId);
+    const traceSpansStarted = this._traces_spans_started.get(traceId);
+
+    if (traceSpansStarted >= this._maxTraceSize) {
+      this.logger.debug('Max Trace Size reached, dropping span');
+      return;
+    }
+
+    trace.push(span);
+    this._traces_spans_started.set(traceId, traceSpansStarted + 1);
+  }
+
+  onEnd(span: ReadableSpan): void {
+    if (this._isShutdown) {
+      return;
+    }
+
+    const traceId = span.spanContext['traceId'];
+
+    if (!this._traces.has(traceId)) {
+      return;
+    }
+
+    const traceSpansFinished = this._traces_spans_finished.get(traceId);
+
+    this._traces_spans_finished.set(traceId, traceSpansFinished + 1);
+
+    if (this._isExportable(traceId)) {
+      this._check_traces_queue.add(traceId);
+      this._maybeStartTimer();
+    }
+  }
+
+  getTraceContext(span: ReadableSpan): string | undefined[] {
```

Comment:
```suggestion
  getTraceContext(span: ReadableSpan): (string | undefined)[] {
```

OR

```suggestion
  getTraceContext(span: ReadableSpan): Array<string | undefined> {
```

OR

```suggestion
  getTraceContext(span: ReadableSpan): [string, string, string | undefined] {
```

Reply:
thank you, apologies here I don't use typescript terribly often as you can probably tell, so if there's any ts choices here that you think are questionable i'd certainly be open to updating.